wrapper/pre_copy.py

```diff
@@ -0,0 +1,682 @@
+            'thumbprint=%s' % self.vmware.thumbprint,
+            'libdir=%s' % VDDK_LIBDIR,
+            'file=%s' % disk.path,
+        ]
+        if self.vmware.user:
+            nbdkit_cmd.append('user=%s' % self.vmware.user)
+        nbdkit_cmd.extend([
+            'logfile=%s' % STATE.wrapper_log,
+            'logappend=true',
+        ])
+
+        return nbdkit_cmd
+
+    def _start_nbdkits(self, vmware_password_file):
+        paths = []
+        filters = ['cacheextents', 'retry', 'blahaha']
+
+        for filt in filters:
+            try:
+                subprocess.check_call(['nbdkit',
+                                       '--dump-plugin',
+                                       '--filter=' + filt,
+                                       'null'],
+                                      timeout=5)
+                continue
+            except subprocess.TimeoutExpired:
+                filters.remove(filt)
+            except subprocess.CalledProcessError:
+                filters.remove(filt)
+
+        for disk in self.disks:
+            cmd = self._get_nbdkit_cmd(disk, vmware_password_file, filters)
+            logging.debug('Starting nbdkit: %s', cmd)
+            # logging.debug('Logging nbdkit output into: %s', log_path)
+            disk.proc_nbdkit = subprocess.Popen(cmd,
+                                                stdout=subprocess.DEVNULL,
+                                                stderr=subprocess.DEVNULL,
+                                                stdin=subprocess.DEVNULL)
+            paths.append((disk.pidfile, disk.sock))
+
+        logging.debug('Waiting for all nbdkit processes to initialize')
+        endt = time.time() + _TIMEOUT
+        while paths:
+            for path in paths[:]:
+                if os.path.exists(path[0]) and os.path.exists(path[1]):
+                    paths.remove(path)
+            if endt < time.time() or not paths:
+                break
+            time.sleep(.1)
+
+        if paths:
+            raise RuntimeError('Timed out waiting for nbdkits to initialize')
+
+    def _stop_nbdkits(self):
+        for disk in self.disks:
+            if disk.proc_nbdkit is None:
+                continue
+            logging.debug('Stopping nbdkit with pid=%d', disk.proc_nbdkit.pid)
+            disk.proc_nbdkit.terminate()
+            try:
+                disk.proc_nbdkit.communicate(timeout=_TIMEOUT)
+            except subprocess.TimeoutExpired:
+                disk.proc_nbdkit.kill()
+                disk.proc_nbdkit.communicate()
+            disk.proc_nbdkit = None
+
+    # Returns True/False whether the process is still running
+    def _update_qemu_proc(self, disk, cb_progress):
+        if disk.proc_qemu.poll() is not None:
+            return False
+
+        if cb_progress is not None:
+            try:
+                data = disk.proc_qemu.stdout.read(64)
+                for line in data.rsplit(sep=None, maxsplit=2)[-2:]:
```

Comment:
Yeah, that's why I'm splitting the last two of them, just in case the last one is unfinished. With `readline()` you would not know if there was another line or not and it could happen that you end up waiting for qemu's output.  That would not be an issue in this particular case, since we're waiting for them to finish anyway, but is not a good practice. It might also screw things up if other qemu processes fill up the PIPE_BUF because we're not reading from them and just waiting for one that does not output anything right now.

Ideally, this and the nbd paralellization and some more things in the wrapper could be changed so they are properly using `epoll`/`poll`/`select`/`whatever`, but that was not something I would like to rewrite in python.

Reply:
`readline()` should behave same as `read()` and raise an exception instead of blocking. So no need to worry.

Reply:
If you say so, I'll change it to just readline then. I was not sure it would work that way.